### ISSUE TITLE
Allow module augmentation for CustomFormatConfig.format

### ISSUE THREAD

**Issue by ImADrafter** [formatjs/formatjs#2709]
## Which package?
@formatjs/intl

## Describe the bug

It would improve DX a lot if `CustomFormatConfig` could be augmented, to contain user defined formats. But, ATM, format is declared as "string | undefined".

```tsx
export interface CustomFormatConfig {
    format?: string;
}
```

So if you try module augmentation over this key, typescript complains since union literal are more specific than string:

```tsx
declare module '@formatjs/intl' {
    export interface CustomFormatConfig {
        format?: 'myNumber' | 'other'; 
        ^^^^^^
        // Subsequent property declarations must have the same type.  Property 'format' must be of type 'string | undefined', but here has type '"myNumber" | "other" | undefined'.
    }
}
```

Module augmentation is used by other popular libraries, like [Emotion](https://emotion.sh/docs/typescript#define-a-theme) and [react-redux](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react-redux/index.d.ts#L46) for example.

This would make the prop "format" more type-safe, only allowing existing formats:

```tsx
    {formatNumber(1000.123213123123, { format: 'non-existing-format' })}
     //  'non-existing-format' is not assignable to 'myUserDefinedFormat'
```

## Expected behavior
`format` should allow use only the user-defined formats.

## Desktop (please complete the following information):

- react-intl@5.13.2
- @formatjs/intl@1.8.2

**Comment by longlho** [comment]
PR's always welcome

**Comment by ImADrafter** [comment]
Sure thing @longlho ; I tried to but I couldn't manage to reach anything successful, so Im asking for some help :). What I've tried so far is:
1) The module augmentation path is going nowhere, since typescript wont allow you to be more specific, just allows you to "expand" the already declared key.
I tried to modify that key type to `any`, with the same results. `any` is still a supertype, so merging wont apply. Maybe we should change that type to a different one ?
2) I tried to pass down a new generic, but, this won't solve the problem, since keys wont be treated as _readonly_ strings (literal strings) if they're not casted `as const`, and it would override the type.

So, if anyone here with more knowledge on Typescript and this library could help me, I would appreciate it.
